### PR TITLE
Remove 'opendatahub.io/link' dependency for notebook controller

### LINKED ISSUE #844: [Feature Request]: Remove `opendatahub.io/link` cases
### Feature description

In the KFNBC flows we need to reply completely on the route object and no longer on the annotation we add.

### Describe alternatives you've considered

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by DaoDaoNoCode**
Closes #844 

## Description

Currently, the notebook controller is highly depending on the `opendatahub.io/link` annotation to get the link of the Jupyter Notebook, this PR gets rid of that. Instead of using that annotation, we use `getRoute` to fetch the route of the notebook as the source of truth.

## How Has This Been Tested?

1. Search in the code base, and make sure there is no `opendatahub.io/link` anymore
2. Try to spawn a notebook using the Jupyter tile
3. Try to access the notebook
4. Try to log out of the notebook
5. Make sure everything works OK

## Test Impact

Add unit tests to test the `useNotebookRedirectLink` hook.

## Request review criteria:

Self checklist (all need to be checked):
- [x] The developer has manually tested the changes and verified that the changes work
- [x] Commits have been squashed into descriptive, self-contained units of work (e.g. 'WIP' and 'Implements feedback' style messages have been removed)
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has added tests or explained why testing cannot be added (unit tests & storybook for related changes)

If you have UI changes: 

- [ ] Included any necessary screenshots or gifs if it was a UI change.
- [ ] Included tags to the UX team if it was a UI/UX change (find relevant UX in the [SMEs](https://github.com/opendatahub-io/odh-dashboard/tree/main/docs/smes.md) section).

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

### CODE REVIEW

**Review by manaswinidas — APPROVED** [review]
/lgtm

Tested. Works fine on my end.

**Comment by lucferbux** [comment]
@DaoDaoNoCode i see you are targeting `f/notebook-controller` here, should we add this to https://github.com/opendatahub-io/odh-dashboard/issues/1845 since we are using that feature branch to get this into incubation

**Comment by DaoDaoNoCode** [comment]
@lucferbux Sure, I think this is fair. Let's add it in, are you going to do this?

**Comment by lucferbux** [comment]
Yes, give me a sec, I'll give this a review tomorrow then!

**Comment by christianvogt** [comment]
lgtm

@DaoDaoNoCode please add a unit test

**Comment by DaoDaoNoCode** [comment]
@christianvogt Can you give me some hints on what kind of unit test I should add for this PR?

**Comment by christianvogt** [comment]
@DaoDaoNoCode we want to make sure we're increasing our test coverage. Easiest tests would be utility changes.

File: backend/src/routes/api/notebooks/utils.ts [hunk 1]
```diff
@@ -17,29 +15,19 @@ export const getNotebookStatus = async (
...
-  }
-
-  return { notebook: newNotebook || notebook, isRunning, podUID };
+  const route = await getRoute(fastify, namespace, name);
+  return {
+    notebook,
+    isRunning,
+    podUID,
+    notebookLink: `https://${route.spec.host}/notebook/${namespace}/${name}`,
+  };
```
Comment (christianvogt) [inline backend/src/routes/api/notebooks/utils.ts:30]: No case handling if retrieving the route fails? Although I don't think i was able to delete the route to have this fetch fail...

Comment (DaoDaoNoCode) [reply backend/src/routes/api/notebooks/utils.ts:30]: Yeah... Maybe only the failed network call can lead to this... let me check, but I believe the operator will always make sure the route is there.

**Comment by DaoDaoNoCode** [comment]
With @christianvogt's help, the test cases are added, it's really for review now.

**Pull request merged**